src/util/datablock/datablock.c

```diff
@@ -0,0 +1,166 @@
+
+    dataBlock->itemCap = dataBlock->blockCount * BLOCK_CAP;
+}
+
+DataBlock *DataBlock_New(size_t itemCap, size_t itemSize) {
+    DataBlock *dataBlock = malloc(sizeof(DataBlock));    
+    dataBlock->itemCount = 0;
+    dataBlock->itemSize = itemSize;
+    dataBlock->blockCount = 0;
+    dataBlock->blocks = NULL;
+    _DataBlock_AddBlocks(dataBlock, ITEM_COUNT_TO_BLOCK_COUNT(itemCap));
+    return dataBlock;
+}
+
+DataBlockIterator *DataBlock_Scan(const DataBlock *dataBlock) {
+    assert(dataBlock);
+    Block *startBlock = dataBlock->blocks[0];
+    return DataBlockIterator_New(startBlock, 0, dataBlock->itemCount, 1);
+}
+
+void *DataBlock_GetItem(const DataBlock *dataBlock, size_t idx) {
+    assert(dataBlock && idx >= 0 && idx < dataBlock->itemCount);
+    Block *block = GET_ITEM_BLOCK(dataBlock, idx);
+    idx = ITEM_POSITION_WITHIN_BLOCK(idx);
+    return block->data + (idx * block->itemSize);
+}
+
+void DataBlock_SetItem(DataBlock *dataBlock, void *item, size_t idx) {
+    assert(dataBlock && idx <= dataBlock->itemCount);
+
+    // Get block.
+    size_t blockIdx = ITEM_INDEX_TO_BLOCK_INDEX(idx);
+    Block *block = dataBlock->blocks[blockIdx];
+
+    // Add item to block.
+    idx = ITEM_POSITION_WITHIN_BLOCK(idx);
+    size_t offset = idx * block->itemSize;
+    memcpy(block->data + offset , item, dataBlock->itemSize);
+}
+
+void DataBlock_AddItems(DataBlock *dataBlock, size_t itemCount, DataBlockIterator **it) {
+    assert(dataBlock && itemCount > 0);
+    
+    // Make sure we've got room for items.
+    if(dataBlock->itemCount + itemCount >= dataBlock->itemCap) {
+        // Allocate twice as much items then we currently hold.
+        size_t newCap = (dataBlock->itemCount + itemCount) * 2;
+        size_t requiredAdditionalBlocks = ITEM_COUNT_TO_BLOCK_COUNT(newCap) - dataBlock->blockCount;
+       _DataBlock_AddBlocks(dataBlock, requiredAdditionalBlocks);
+    }
+
+    if(it) {
+        int step = 1;
+        Block *block = ACTIVE_BLOCK(dataBlock);
+        *it = DataBlockIterator_New(block, dataBlock->itemCount,
+                                    dataBlock->itemCount + itemCount, step);
+    }
+
+    dataBlock->itemCount += itemCount;
+}
+
+Block *DataBlock_GetItemBlock(const DataBlock *dataBlock, size_t itemIdx) {
+    return GET_ITEM_BLOCK(dataBlock, itemIdx);
+}
+
+void DataBlock_MigrateItem(DataBlock *dataBlock, size_t src, size_t dest) {
+    assert(dataBlock);
+
+    // Get the block in which dest item resides.
+    Block *destItemBlock = GET_ITEM_BLOCK(dataBlock, dest);
+    
+    // Get item position within its block.
+    size_t destItemBlockIdx = ITEM_POSITION_WITHIN_BLOCK(dest);
+    size_t destItemOffset = (destItemBlockIdx * dataBlock->itemSize);
+
+    // Get the src item.
+    void *srcItem = DataBlock_GetItem(dataBlock, src);
+
+    // Replace dest node with src node.
+    memcpy(destItemBlock->data + destItemOffset, srcItem, dataBlock->itemSize);
+
+    dataBlock->itemCount--;
```

Comment:
This logic works correctly given the places from which we invoke `DataBlock_MigrateItem` right now, but it makes me a little nervous.

Specifically, it accepts an arbitrary source element, but always invalidates the last element rather than the source element. (Decrementing itemCount will cause the last element to get written over on the next addition, as well as triggering matrix resizes that will remove the last row/column).

This works right now because this function is only used from the delete routine, so the source will either be the last element or the last element will be a deletion candidate anyway (at least, this will be true following the merge of #99). To be valid for non-last source elements, though, we'd need to memcpy the last element to the source's position, which would then necessitate matrix updates.